---
title: "RNA-seq on India data--incomplete"
author: "W. Evan Johnson"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    theme: "flatly"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse) ## Data management, plotting (ggplot2), and summaries
library(SummarizedExperiment) ## Data management and storage
library(DT) ## Nice tables and tabular displays (datatable function)
library(gridExtra) ## Displaying two or more plots together
library(umap) ## Calculating the umap for batch correction and summaries
library(sva) ## Batch correction via ComBat
library(DESeq2) ## Differential expression 
library(ComplexHeatmap) ## Making beautiful heapmaps (Heatmap function)
library(TBSignatureProfiler) ## Profiling existing TB signatures in data
```

## RNA-seq analysis on the Indian Dataset {.tabset}

### Load in Dataset
Read in the `indata.rds` dataset. Which assays are availble in this dataset. Print out the sample information data (colData). 

```{r}
indata <- readRDS("~/Desktop/indata.rds") ## change directory information

### lists the availble gene expression types
assays(indata)

### gene expression data
counts <- assay(indata, "counts")
#head(counts)

### sample information
colData(indata)
```

### Summarize the colData (Table 1 in paper) {.tabset}
Summarize the sample information data, provide relevant tables and plots. 

#### Save colData as a Tibble
```{r}
column_dat <- as_tibble(colData(indata))
datatable(column_dat)
```

#### Summarize TB Status
```{r}

```

#### Summarize BMI by TB Status 
```{r}

```

#### Summarize Malnourishment Status
```{r}

```

#### Summarize Malnourishment by TB Status
```{r}

```

### Batch effects correction {.tabset}
Evaluate whether batch effects are present in the data and if correction if is needed. If so, correct for batch. Use the `log_counts_cpm` assay for plotting.

#### PCA plot (unadjusted)
Plot the first two principal components, colored by batch, TB status as symbols 

```{r}
log_cpm_dat <- as.matrix(assay(indata, "log_counts_cpm"))

```

#### UMAP plot (unadjusted)
Plot the UMAP projection, colored by batch, TB status as symbols 
```{r}

```

#### Batch correction (ComBat and ComBat-seq)
Adjust the `log_counts_cpm` with ComBat and the `counts` using ComBat-Seq. Use `Tb_status` and `bmi_cat2` as covariates. 
```{r}
## ComBat on log_counts_cpm
cov_model <- model.matrix(~Tb_status+bmi_cat2,
                          data=colData(indata))
combat_adjust <- ComBat(log_cpm_dat, colData(indata)$batch, cov_model)
assay(indata,"combat_log_counts_cpm") <- combat_adjust

## ComBat-Seq on counts

```

#### PCA plot (batch adjusted)
Plot the first two principal components, colored by batch, TB status as symbols 

```{r}

```

#### UMAP plot (batch adjusted)
Plot the UMAP projection, colored by batch, TB status as symbols 
```{r}
set.seed(0)

```


### Conduct differential expression (TB vs LTBI) {.tabset}
Use DESeq2 to calculate differential expression using the ComBat-Seq adjusted counts between TB and LTBI (`Tb_status`), adjusting for malnutrition status (`bmi_cat2`).

### Apply DESeq2 
```{r}

``` 

#### Heatmap of DEGs

```{r}
# Make a Heatmap of DEGs

```


### TBSignatureProfiler Analysis {.tabset}
The TBSignatureProfiler was developed in the Johnson Lab in 2021 to profile new and existing TB gene expression signatures [https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-020-05598-z](https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-020-05598-z). Help documentation and vignettes are available at [https://wejlab.github.io/TBSignatureProfiler-docs/](https://wejlab.github.io/TBSignatureProfiler-docs/])

```{r, message=FALSE, results='hide'}
TBsigs <- TBsignatures[-12] ## Remove Chendi_HIV_2 signature because we only have one of the genes
ssgsea_res <- runTBsigProfiler(indata, useAssay = "combat_log_counts_cpm",
                                 signatures = TBsigs,
                                 algorithm = "ssGSEA",
                                 combineSigAndAlgorithm = TRUE,
                                 parallel.sz = 1)
```

#### Signature Heatmap

```{r}

```

#### Signature Boxplots

```{r , fig.width=12, fig.height=9}

```

#### Boxplots Single {.tabset}

```{r , results="asis"}
for (i in names(TBsigs)){

  cat("#####" , i, "\n")

  print(signatureBoxplot(ssgsea_res, name=i, signatureColNames = i,
                 annotationColName = c("Tb_status")))
  
  cat("\n\n")
}

```

#### Signature Gene Heatmaps {.tabset}

```{r , results="asis"}
for (i in names(TBsigs)){

  cat("#####" , i, "\n")

  
  
  cat("\n\n")
}
```

#### AUC Table
```{r, message = FALSE}
set.seed(0)

```

#### AUC Boxplots

```{r, message = FALSE}
set.seed(0)


```

#### ROC plots
 
```{r, message = FALSE, fig.height = 9, fig.width = 12}


```

#### Separate ROC plots {.tabset}

```{r, results = 'asis', message = FALSE}
for (i in names(TBsigs)){
  
  cat("#####" , i, "\n")
  
 
  
  
  cat("\n\n")
}
```

