---
title: "RNA-seq on India data"
author: "W. Evan Johnson"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    theme: "flatly"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(SummarizedExperiment)
library(DT)
library(gridExtra)
library(umap)
library(sva)
library(DESeq2)
library(ComplexHeatmap)
library(TBSignatureProfiler)
```

## RNA-seq analysis on the Indian Dataset {.tabset}

### Load in Dataset
Read in the `indata.rds` dataset. Which assays are availble in this dataset. Print out the sample information data (colData). 

```{r}
indata <- readRDS("~/Desktop/indata.rds")

### lists the availble gene expression types
assays(indata)

### gene expression data
counts <- assay(indata, "counts")
#head(counts)

### sample information
colData(indata)
```

### Summarize the colData (Table 1 in paper) {.tabset}
Summarize the sample information data, provide relevant tables and plots. 

#### Save colData as a Tibble
```{r}
column_dat <- as_tibble(colData(indata))
datatable(column_dat)
```

#### Summarize TB Status
```{r}
column_dat %>% group_by(Tb_status) %>% 
  summarize (n = n()) %>%
  mutate(freq = n / sum(n))

column_dat %>% 
  ggplot(aes(Tb_status,fill=Tb_status)) +
  geom_bar()
```

#### Summarize BMI by TB Status 
```{r}
column_dat %>% group_by(Tb_status) %>% 
  summarize(mean(BMI))

column_dat %>% 
  ggplot(aes(BMI,fill=Tb_status)) +
  geom_density(alpha=0.2)
```

#### Summarize Malnourishment Status
```{r}
column_dat %>% group_by(bmi_cat2) %>% 
  summarize (n = n()) %>%
  mutate(freq = n / sum(n))

column_dat %>% 
  ggplot(aes(bmi_cat2,fill=bmi_cat2)) +
  geom_bar()
```

#### Summarize Malnourishment by TB Status
```{r}
column_dat %>% group_by(bmi_cat2, Tb_status) %>% 
  summarize (n = n()) %>%
  mutate(freq = n / sum(n))

p1 <- column_dat %>% ggplot(aes(bmi_cat2,fill=Tb_status)) +
  geom_bar() + ggtitle("Malnourishment by TB Status")

p2 <- column_dat %>% ggplot(aes(Tb_status,fill=bmi_cat2)) +
  geom_bar() + ggtitle("TB Status by Malnourishment")

grid.arrange(p1, p2, nrow=1)
```

### Batch effects correction {.tabset}
Evaluate whether batch effects are present in the data and if correction if is needed. If so, correct for batch. Use the `log_counts_cpm` assay for plotting.

#### PCA plot (unadjusted)
Plot the first two principal components, colored by batch, TB status as symbols 

```{r}
log_cpm_dat <- as.matrix(assay(indata, "log_counts_cpm"))
pcas <- as.data.frame(prcomp(t(log_cpm_dat))$x)
pcas$batch <- colData(indata)$batch
pcas$Tb_status <- colData(indata)$Tb_status

pcas %>% ggplot(aes(PC1, PC2, col=batch, pch = Tb_status)) +
  geom_point()
```

#### UMAP plot (unadjusted)
Plot the UMAP projection, colored by batch, TB status as symbols 
```{r}
set.seed(0)
umaps <- as.data.frame(umap(t(log_cpm_dat))$layout)
umaps$batch <- colData(indata)$batch
umaps$Tb_status <- colData(indata)$Tb_status

umaps %>% ggplot(aes(V1, V2, col=batch, pch = Tb_status)) +
  geom_point()
```

#### Batch correction (ComBat and ComBat-seq)
Adjust the `log_counts_cpm` with ComBat and the `counts` using ComBat-Seq. Use `Tb_status` and `bmi_cat2` as covariates. 
```{r}
## ComBat on log_counts_cpm
cov_model <- model.matrix(~Tb_status+bmi_cat2,
                          data=colData(indata))
combat_adjust <- ComBat(log_cpm_dat, colData(indata)$batch, cov_model)
assay(indata,"combat_log_counts_cpm") <- combat_adjust

## ComBat-Seq on counts
counts <- assay(indata, "counts")
combat_seq_adjust <- ComBat_seq(counts, colData(indata)$batch, covar_mod = cov_model)
assay(indata,"combatseq_counts") <- combat_adjust
```

#### PCA plot (batch adjusted)
Plot the first two principal components, colored by batch, TB status as symbols 

```{r}
combat_log_cpm_dat <- as.matrix(assay(indata, "combat_log_counts_cpm"))
pcas <- as.data.frame(prcomp(t(combat_log_cpm_dat))$x)
pcas$batch <- colData(indata)$batch
pcas$Tb_status <- colData(indata)$Tb_status

pcas %>% ggplot(aes(PC1, PC2, col=batch, pch = Tb_status)) +
  geom_point()
```

#### UMAP plot (batch adjusted)
Plot the UMAP projection, colored by batch, TB status as symbols 
```{r}
set.seed(0)
umaps <- as.data.frame(umap(t(combat_log_cpm_dat))$layout)
umaps$batch <- colData(indata)$batch
umaps$Tb_status <- colData(indata)$Tb_status

umaps %>% ggplot(aes(V1, V2, col=batch, pch = Tb_status)) +
  geom_point()
```


### Conduct differential expression (TB vs LTBI) {.tabset}
Use DESeq2 to calculate differential expression using the ComBat-Seq adjusted counts between TB and LTBI (`Tb_status`), adjusting for malnutrition status (`bmi_cat2`).

### Apply DESeq2 
```{r}
dds <- DESeqDataSetFromMatrix(countData = combat_seq_adjust, 
                              colData=colData(indata), 
                              design=~Tb_status+bmi_cat2)

dds<-dds[rowSums(counts(dds))>1,] 
dds<-DESeq(dds) 
res <- as.data.frame(results(dds)) %>% arrange(padj)
rownames(res)

datatable(res)
``` 

#### Heatmap of DEGs

```{r}
# Make a Heatmap of DEGs
mat = as.matrix(assay(indata,"combat_log_counts_cpm"))[
  order(results(dds)[,6])[1:100],] 
mat = t(scale(t(mat)))

df=data.frame("TB_Status"=colData(indata)$Tb_status,
              "Malnutrition"=colData(indata)$bmi_cat2) 

ha = HeatmapAnnotation(df = df, 
                       col = list(TB_Status=c("TB"="red","LTBI"="blue"),
                                  Malnutrition=c("mal"="purple","well"="green")
                                  )
                       )

Heatmap(mat,show_row_names=F,show_column_names = F, top_annotation = ha)
```


### TBSignatureProfiler Analysis {.tabset}
The TBSignatureProfiler was developed in the Johnson Lab in 2021 to profile new and existing TB gene expression signatures [https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-020-05598-z](https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-020-05598-z). Help documentation and vignettes are available at [https://wejlab.github.io/TBSignatureProfiler-docs/](https://wejlab.github.io/TBSignatureProfiler-docs/])

```{r, message=FALSE, results='hide'}
TBsigs <- TBsignatures[-12] ## Remove Chendi_HIV_2 signature because we only have one of the genes
ssgsea_res <- runTBsigProfiler(indata, useAssay = "combat_log_counts_cpm",
                                 signatures = TBsigs,
                                 algorithm = "ssGSEA",
                                 combineSigAndAlgorithm = TRUE,
                                 parallel.sz = 1)
```

#### Signature Heatmap

```{r}
# Colors for gradient
signatureHeatmap(ssgsea_res, name = "Heatmap of Signatures (ssGSEA)", 
                 signatureColNames = names(TBsigs),
                 annotationColNames = c("Tb_status"),
                 scale = TRUE,
                 split_heatmap = "none",
                 showColumnNames = FALSE)
```

#### Signature Boxplots

```{r , fig.width=12, fig.height=9}
signatureBoxplot(ssgsea_res, name="ssGSEA", 
                 signatureColNames = names(TBsigs),
                 annotationColName = c("Tb_status"))
```

#### Boxplots Single {.tabset}

```{r , results="asis"}
for (i in names(TBsigs)){

  cat("#####" , i, "\n")

  print(signatureBoxplot(ssgsea_res, name=i, signatureColNames = i,
                 annotationColName = c("Tb_status")))
  
  cat("\n\n")
}

```

#### Signature Gene Heatmaps {.tabset}

```{r , results="asis"}
for (i in names(TBsigs)){

  cat("#####" , i, "\n")

  signatureGeneHeatmap(ssgsea_res, useAssay="log_counts_cpm", 
                     TBsigs[[i]],
                     name = i, signatureColNames = NULL, 
                    annotationColNames = c(i,"Tb_status"),
                     showColumnNames = FALSE)
  
  cat("\n\n")
}
```

#### AUC Table
```{r, message = FALSE}
set.seed(0)
tableAUC(ssgsea_res, 
         annotationColName = "Tb_status", 
         signatureColNames = names(TBsigs),
         num.boot = 100, 
         pb.show = FALSE)
```

#### AUC Boxplots

```{r, message = FALSE}
set.seed(0)
compareBoxplots(ssgsea_res, annotationColName = "Tb_status",
                signatureColNames = names(TBsigs),
                pb.show = FALSE, fill.col = "blue",
                rotateLabels = TRUE)
```

#### ROC plots
 
```{r, message = FALSE, fig.height = 9, fig.width = 12}
signatureROCplot_CI(inputData = ssgsea_res,
                   signatureColNames = names(TBsigs),
                   annotationColName = "Tb_status")

```

#### Separate ROC plots {.tabset}

```{r, results = 'asis', message = FALSE}
for (i in names(TBsigs)){
  
  cat("#####" , i, "\n")
  
  print(signatureROCplot_CI(inputData = ssgsea_res, 
                   signatureColNames = i,
                   annotationColName = "Tb_status",
                   name = paste("ROC plot,", i, sep = " ")))
  
  cat("\n\n")
}
```

